Proofs/Experiments/AddLoop.lean

```diff
@@ -0,0 +1,655 @@
+  (h_sp_aligned : CheckSPAlignment s)
+  (h_step : sn = run 1 s) :
+  cut sn = true ∧
+  r StateField.PC sn = 0x4005bc#64 ∧
+  r StateField.ERR sn = .None ∧
+  sn.program = program ∧
+  CheckSPAlignment sn := by
+  have := program.stepi_0x4005b8 s (run 1 s) h_program h_pc h_err
+  simp only [minimal_theory] at this
+  simp_all only [run, cut, this, state_simp_rules, bitvec_rules, minimal_theory]
+  done
+
+-- (TODO) program.stepi_0x4005bc_cut elided.
+
+-------------------------------------------------------------------------------
+
+private theorem AddWithCarry.all_ones_identity_64 (x : BitVec 64) :
+  (AddWithCarry x 0xffffffffffffffff#64 0x1#1).fst = x := by
+  simp only [AddWithCarry]
+  bv_decide
+
+private theorem AddWithCarry.all_ones_zero_flag_64 (x : BitVec 64) :
+  ((AddWithCarry x 0xffffffffffffffff#64 0x1#1).snd.z = 1#1) ↔
+  (x = 0#64) := by
+  simp only [AddWithCarry, bitvec_rules, state_simp_rules]
+  repeat split
+  · bv_decide
+  · bv_decide
+  done
+
+theorem nextc_to_run_from_0x4005a4 (h : pre s0) :
+  (Correctness.nextc (Sys.run s0 1)) = run 2 s0 := by
+  simp only [pre, state_simp_rules] at h
+  obtain ⟨h_s0_pc, h_s0_program, h_s0_err⟩ := h
+  simp only
+    [Correctness.nextc, Correctness.arm_run,
+     Spec'.cut, minimal_theory]
+
+  rw [AddLoop.csteps_eq]
+  have h_step_1 := program.stepi_0x4005a4_cut s0 (run 1 s0)
+  simp_all only [minimal_theory, bitvec_rules]
+  generalize h_s1 : run 1 s0 = s1 at h_step_1
+
+  -- Stop when the *_cut theorem says (cut <state>) = true.
+  rw [AddLoop.csteps_eq]
+  have h_step_2 := program.stepi_0x4005b0_cut s1 (run 1 s1)
+  simp_all only [minimal_theory]
+  generalize h_s2 : run 1 s1 = s2 at h_step_2
+
+  simp only [←h_s2, ←h_s1, ←run_plus]
+  done
+
+theorem effects_of_nextc_from_0x4005a4
+  (h_pre : pre s0)
+  (_h_pc : r StateField.PC s0 = 0x4005a4#64)
+  (h_run : sf = (Correctness.nextc (Sys.run s0 1))) :
+  -- I've elided irrelevant effects and non-effects for now.
+  sf.program = program ∧
+  r StateField.PC sf = 0x4005b4#64 ∧
+  r StateField.ERR sf = .None ∧
+  ((r (StateField.FLAG PFlag.Z) sf = 1#1) ↔
+   (r (StateField.GPR 0x0#5) s0 = 0x0#64)) ∧
+  r (StateField.GPR 0#5) sf = r (StateField.GPR 0#5) s0 ∧
+  r (StateField.GPR 1#5) sf = r (StateField.GPR 1#5) s0 ∧
+  CheckSPAlignment sf := by
+  -- Prelude
+  rw [nextc_to_run_from_0x4005a4 h_pre] at h_run
+  -- TODO: Tactic to "explode" conjunctions?
+  obtain ⟨h_s0_pc, h_s0_program, h_s0_err, h_s0_sp_aligned⟩ := h_pre
+  -- Symbolic simulation
+  -- (FIXME) sym1_n doesn't play well with unconditional branches.
+  /-
+  application type mismatch
+  program.stepi_0x4005a8 s1 s2 h_s1_program h_s1_pc
+  argument
+  h_s1_pc
+  has type
+  r StateField.PC s1 = 0x4005b0#64 : Prop
+  but is expected to have type
+  r StateField.PC s1 = 0x4005a8#64 : Prop
+  -/
+  -- sym1_n 2
+  sym1_n 1 at s0
+  sym1_n 1 at s1
+  -- (FIXME) better stepi lemma generation
+  simp (config := {ground := true}) only at h_step_2
```

Comment:
What, concretely, needs improving here?

Reply:
Badly placed comment there. `sym1_n 1 at s1` (the previous tactic) generates proof state with subterms like:
```
(AddWithCarry (r (StateField.GPR 0x0#5) s1)
          (Decidable.rec (fun h => (0x0#1, 0x0#64)) (fun h => (0x1#1, 0xffffffffffffffff#64))
              (instDecidableEqBitVec 0x1#1 0x1#1)).snd
          (Decidable.rec (fun h => (0x0#1, 0x0#64)) (fun h => (0x1#1, 0xffffffffffffffff#64))
              (instDecidableEqBitVec 0x1#1 0x1#1)).fst)
```
which go away with the `simp/ground`. In an ideal world, we shouldn't need this `simp/ground`.